torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -109,6 +109,17 @@ ProcessGroupAgent::ProcessGroupAgent(
   }
 }
 
+ProcessGroupAgent::~ProcessGroupAgent() {
+  try {
```

Comment:
It's not completely clear to me that this logic belongs in this change... if we really hit the destructor, and it throws, it seems that we might want to just let it throw (same in ThriftRpcAgent)

I totally see why we watch to catch if a deserialization throws and fails, but this is always an exit case, and not entirely clear how we want to respond to that. In such as case, it can be a lesser evil to just keep the code simple and punt.

Reply:
This was from a comment I got on the previous version of this PR that I linked, it seems that in general, we want to avoid throwing in C++ destructors?

Reply:
Right, throwing in c++ destructors is generally discouraged.
But it's not particularly "dangerous" here - this basically only happens at the end of process (i.e. not like leaking memory matters), and swallowing exceptions here might hide useful information.

If others disagree, I don't hold the view that strongly, but do probably think the destructor part is unnecessary here, at this point.

---
Context (which might not be appropriate here): In some prod services that I've worked on, the philosophy was that if we didn't understand a weird failure case, or it wasn't clear how to deal with it, especially if it was rare - it can made sense to just allow the process to LOG(FATAL), with useful debug info (like the thrown exception here). When it actually happens, the team then has more data to figure out if/how to handle the case.

But when you try to error recover or swallow errors too aggressively, you both make the code larger/harder to understand, and also sometimes make it harder to notice the malfunction.

Note that this philosophy definitely doesn't apply in all contexts (e.g. it's very bad in Mobile or client-side shipped software), but in case of a new distributed framework that we're bringing up, and have the ability to update, it might be appropriate to not swallow exceptions too aggressively.

Reply:
Thanks, that's really helpful advice. I do think we need the destructor itself, otherwise in the test, at least under `spawn` mode, an error is thrown complaining that the threads in `process_group_agent` weren't terminated properly. However I think we can do without the try/catch. @pritamdamania87 do you have any suggestions around the destructor and whether we should catch exceptions in it or not?

Reply:
So we are seeing this in real life? I think the consequences of this aspect of the change are more unpredictable than catching errors in the send thread.

Reply:
@jjlilley Are you referring to the destructor itself or the try/catch in the destructor? I don't think we're seeing the issue I described in the previous comment in any tests or uses I've seen, but this is because the user is always able to call `rpc.shutdown()`. Without the destructor at all, I'm getting an error `terminate called without an active exception`in the test I added, which seems to be because the threads are not terminated gracefully. If a similar situation can happen in a real use case, wouldn't it be useful to at least attempt to `shutdown` in the destructor?

Reply:
Sorry for the ambiguity. shudown() in the destructor seems quite reasonable.
Catching the error feels like it has the potential to hide something.